### DISCUSSION TITLE
Dynamically updating Runtime Caching Strategies (RCS)

### DISCUSSION THREAD

**Discussion by AlvSovereign** [serwist/serwist#61] [Q&A] [answered]
Hi all

How would you come up with a solution to have the ability to dynamically update RCS when the Service Worker has already been registered with a different list of RCS?

The use case is that we may want to remove caching entirely, via a feature flag. The feature flag will be set to either allow caching or remove caching for a browser depending on a business scenario, which could change at any time.

The best I have come up with is to:
- in an additional script, `postMessage` to the SW with the updated list of RCSs (whilst having to use a custom stringify replacer to handle functions and regexes) 
- listen to messages in the SW and on receiving the correct `type` of message, we parse (with custom replacer) the updated list of RSCs
- overwrite the `fetch` listener with a new set of RCS handles depending on its options

Does there/or is there an easier, more robust way handle this?

**Comment by DuCanhGH** [comment]
@AlvSovereign this use case seems fairly complex. I'd say that there isn't really a robust way to do it with `runtimeCaching`.

However, you can reimplement this entirely with `@serwist/routing`:
```ts
import type { PrecacheEntry } from "@serwist/precaching";
import { installSerwist } from "@serwist/sw";
import { Router, Route } from "@serwist/routing";

declare const self: ServiceWorkerGlobalScope & {
  // Change this attribute's name to your `injectionPoint`.
  __SW_MANIFEST: (PrecacheEntry | string)[] | undefined;
};

// Note: `@serwist/routing.registerRoute` uses a global Router that is inaccessible to the user. Don't use it. We need
// the `Router.unregisterRoute` method. We will probably implement `@serwist/routing.unregisterRoute` some time soon,
// but not now.
const router = new Router();
// You may look at https://github.com/serwist/serwist/blob/36f780c5fb401cadafb06b77a8cd30cdcb274689/packages/sw/src/registerRuntimeCaching.ts
// to see how `runtimeCaching` is implemented and implement a similar method for convenience.
const routes = [
  new Route(/\.(?:js)$/i, new StaleWhileRevalidate({
    cacheName: "static-js-assets",
    plugins: [
      new ExpirationPlugin({
        maxEntries: 48,
        maxAgeSeconds: 24 * 60 * 60, // 24 hours
      }),
    ],
  })),
  new Route(/\.(?:css|less)$/i, new StaleWhileRevalidate({
    cacheName: "static-style-assets",
    plugins: [
      new ExpirationPlugin({
        maxEntries: 32,
        maxAgeSeconds: 24 * 60 * 60, // 24 hours
      }),
    ],
  })),
];

router.addFetchListener();
router.addCacheListener();

for (const route of routes) {
  router.registerRoute(route);
}

self.addEventListener("message", (event) => {
  if (event.data && event.data.type === "REMOVE_ROUTE") {
    // Note: It is recommended that you use a more readable and robust way of 
    // identifying routes than indexes.
    router.unregisterRoute(routes[event.data.index]);
  }
});

installSerwist({
  precacheEntries: self.__SW_MANIFEST,
  skipWaiting: true,
  clientsClaim: true,
  navigationPreload: true,
  // Note: You may use `runtimeCaching` alongside the above code. They can
  // work together.
  runtimeCaching: undefined,
});
```
Note that we don't try to stringify a list with functions and RegExp's here. Simply define them in the source then dynamically add or remove them with `registerRoute` and `unregisterRoute`.

**Reply by AlvSovereign** [reply, marked as answer]
Thanks for the reply - super helpful.

**Reply by AlvSovereign** [reply]
I've finally had some time to look into this again and found this error during `router.addFecthListeners()`: `Event handler of 'fetch' event must be added on the initial evaluation of worker script.`

It seems that the event handler for the fetch event needs to be registered during the initial execution of the Service Worker script as not to miss any network requests the client may make.

I was trying to initialise the SW without intercepting any fetch request, then setting up the fetch listeners some time in the future via a message event, to hopefully start intercepting requests from that point onwards - basically applying your `router` logic above inside the `message` event listener.

**Reply by DuCanhGH** [reply]
@AlvSovereign you can't conditionally add fetch listeners indeed. I don't see why you'd need to do that though. You can simply call `registerRoute` and `unregisterRoute` conditionally.
